### ISSUE TITLE
iOS Beta, version 2024.25+

### ISSUE THREAD

**Issue by biblicabeebli** [onnela-lab/beiwe-ios#69]
## **2024.45** is live on Testflight and contains a breaking change! Old versions of the app will get stuck in a resend loop. Please upgrade.

- This build has had virtually all code libraries updated due to cascading requirements changes, random things may have changed. We need wide-spectrum testing of all features.
- Please include weekly surveys in your test studies.
  - They don't need to test anything in the survey, there is an issue where a participant could fail to correctly generate new "Scheduled Events" affecting weekly surveys.
- Please report graphical glitches and other visually-observable changes.
  - With screenshots!
- Please attempt installing this version of the app over an existing Beiwe app installation from the app store to ensure we get decent coverage.
- There is a known issue (purely on the backend) where a push notification resend could get stuck in an endless resurrection loop causing infinite resends.
  - There is a possible fix deployed to the staging server - this issue is entirely resolvable on the backend.

#### we can resolve these in a future release of an update to researchkit
- Two survey navigation bugs have been identified:
  - one is a crash involving navigation with the back button.
    - https://github.com/onnela-lab/beiwe-ios/issues/71
  - one is a new and exciting way of triggering a scenario where the user can no longer move through a survey because the next and skip buttons disappear.
    - https://github.com/onnela-lab/beiwe-ios/issues/72

## Notes so far:
- ~~black/white bar during registration.~~
- see https://github.com/onnela-lab/beiwe-ios/issues/59 for discussion
- see https://github.com/onnela-lab/beiwe-backend/issues/402 for backend investigation.
- ~registration-time crash on ios 18?~ The app crashed on iOS 18 if installed through TestFlight (and presumably the App Store, as they are very similar). Underlying cause was some new compiler/os-level security protection that broke a use of a mutable pointer inside the encryption code.
- - Squashed bug: https://github.com/onnela-lab/beiwe-ios/issues/47
-   - The UI libraries were revved, and believe-it-or-not you are probably a better judge of whether something has changed
  - (The white/black bar the registration page has been fixed as of 2024.44.)

**Comment by MMel099** [comment]
I'm having crashing with this new version. Tried installing on top of the old and then uninstalling - getting crashes with both. After uninstalling, I was able to open the app and perform a few actions before it would crash. Originally tried this on iOS 17.5.1 and have since updated to the new 18.0.1. On my new OS, I am unable to open the app.

**Comment by biblicabeebli** [comment]
(perfect....)

**Comment by jprince127** [comment]
Hello! I wanted to concur with Max, I'm unable to open the app due to it crashing. The app either immediately crashes without displaying any graphics or it crashes during the screen with the beiwe logo. I'm on an iPhone with the iOS version 18.0.

**Comment by biblicabeebli** [comment]
@jprince127 @MMel099 can you try build 2024.27?

**Comment by biblicabeebli** [comment]
It is not crashing when launched in the ios 18 simulator....
Were you upgrading over an older app installation?

**Comment by biblicabeebli** [comment]
@jprince127 @MMel099 @hydawo  can you try build 2024.28 and report back?  I will be pushing it out shortly.

**Comment by MMel099** [comment]
@biblicabeebli Hey Eli, is 2024.28 on TestFlight? I'm just seeing 2024.27 on there right now. I just installed that one and got through registration and settings but then it started crashing.

**Comment by biblicabeebli** [comment]
Not yet, sometimes it takes a while.

**Comment by biblicabeebli** [comment]
2024.28 is now available on testflight.

**Comment by MMel099** [comment]
Just saw it come through - downloaded on top of 2024.27 and it crashed immediately. After deleting and redownloading, it let me register and navigate the settings before crashing. Now it's crashing upon start.

**Comment by hydawo** [comment]
Running iOS 18 on iPhone 15 Pro Max using latest TestFlight iOS v2.5.2 (2024.28). I'm testing on my Staging study - [Hassan_Beta_Test_2_13_23](https://staging.beiwe.org/view_study/80) using ID [j8xvmhdl](https://staging.beiwe.org/view_study/80/participant/j8xvmhdl).

Study is configured with:
- Daily survey at 10AM
- Relative surveys at 11AM for Days 1-15 anchored to Enrollment Date
- Absolute surveys (7 surveys scheduled for the week prior to registration) 

- Installed from Testflight -> app crashed on registration
- Launched app -> app crashed
- I did receive a push notification (despite the above crashes), but app crashed when interacting with notification. Push notification page shows resend logic kicked in for all of the past absolute survey times
![IMG_5011](https://github.com/user-attachments/assets/5da7c5f7-bb7a-4eee-987a-6650d6d013fe)

Will wait for next TestFlight version to continue testing...

@MMel099 @jprince127 when the next build is released in TestFlight, can you please register to my study above. You can create a new ID for you to use. 

@biblicabeebli any other testing conditions we should specifically configure the study for?

**Comment by biblicabeebli** [comment]
there's updates about what different releases are tests of in the release notes, I have to push out variations.

Old releases / their notes are still available under the previous releases menu in testflight if you want to check progress.

**Comment by biblicabeebli** [comment]
(I will make it very clear in the app release description when it [finally] works, and post here.)

**Comment by biblicabeebli** [comment]
@jprince127 @hydawo @MMel099 I've pushed out Build 2024.42 and 2024.43
- 2024.42 was the first attempt that finally got iOS 18 working again
- 2024.43 is cleanup after the fact.

I'm keeping 2024.42 up for now just in case, but I'm going to retire everything before them.

**Comment by biblicabeebli** [comment]
New bug!
It looks like some action on the website regenerates absolute survey scheduled events - @hydawo registered an absolute survey, something happened, then received those one-time absolute surveys again.

This is probably unrelated to the new work, it is probably an existing bug.  @biblicabeebli check when we regenerate all schedules, we probably have something regenerating all scheduled events instead of a specific type of scheduled event.

@hydawo will test a repeat of actions on the frontend to see if he can replicate it.

**Comment by biblicabeebli** [comment]
@MMel099 @jprince127 @hydawo  Please watch the app badge counter to make sure it is fixed.

**Comment by hydawo** [comment]
@MMel099 @jprince127 some specific tests to run:

- Daily survey at 10AM
- Relative surveys at 11AM for Days 1-15 anchored to Enrollment Date
- Absolute surveys (7 surveys scheduled for the week prior to registration) --> these should be sent shortly after registration. Take note of 1) how many notifications you receive and 2) after taking the survey, does it repopulate in the survey queue/do you continue to get notifications?

**Max**
What happens if you don’t take a survey?
- Test 1: don't take survey until 7 or 8pm
- Test 2: don't take survey at all for a day, pay attention to notification behavior on day after. Make sure there's only 1 copy of each daily/relative survey to take the day after you skip them


**Jenny**
Do pushes and survey package download occur over data as well as wifi?
- Test 1: take your phone off WiFi for a day, record behavior of notifications

Does the device communicate with the server once a survey is complete?
- For example, if I have a survey to complete daily at 9 and I have push notifications going out every 30 minutes, and I finish at 9:30. Will I get notifications for the rest of the day?
- No, device confirms that survey was received when user interacts with notification or opens app
- Test 1: interact with a survey notification, make sure survey shows in queue, do not take and see if notifications continue
- Test 2: interact with a survey notification, make sure survey shows in queue and start taking survey, do not complete, see if notifications continue

Is resend logic timer relative to individual survey/deployment time or is it "global" across all surveys (all surveys are on same timer cycle)
- Test 1: Receive weekly Survey at 10:15AM, don't take, don't open. Confirm you receive resend push at 10:45AM. Don't take, don't open.
- Receive relative Survey at 11AM, don't take, don't open. 
- 11:15AM How many resend pushes do you receive?
- 11:30AM How many resend pushes do you receive?


**Eli**
**IMPORTANT:** If participant is in different timezone than study and notifications localize to that timezone, does resend logic check against the study timezone deployment time or the participant localized deployment time?

**IMPORTANT** 
TLDR: Is resend logic timer relative to individual survey/deployment time or is it "global" across all surveys (all surveys are on same timer cycle)
- Survey A deploys at 10AM, Survey B deploys at 10:45AM. Resend logic is set to cycle every 30 min.
- User doesn't take either survey when it comes in. At 10:30AM user should get resend push for Survey A
- User receives Survey B at 10:45AM and doesn't take
- What happens at 11AM? Do both Survey A and B send resend pushes or does Survey A resend come at 11AM and Survey B resend at 11:15AM

Won’t more frequent notifications overwhelm the notification dashboard and make it less interpretable for study teams?
- Yes, let's explore having these notifications hidden by default, but can be toggled on if needed (similar to heartbeat)

**TLDR:** what happens if resend logic of instance 1 of Survey A coincides with initial deployment of instance 2 of Survey A
What happens if, say, you have a survey on a daily schedule, but the push notifications go out every 24 hours
Would theoretically two “pings” go out at the exact same time? One for the push and one for the normal survey schedule? Or exactly one? 
What happens if, say, you have a survey on a daily schedule, but the resend logic was accidentally configured to go out every 26 hours
Are there any possible occurrences where someone would end up filling the same survey twice?

**Done**
How does this affect “always available surveys?
- Test 1: Max/Hassan/Jenny took always available survey during RA meeting today (11/5/24). Survey repopulates. No notifications thus far.
What are the resend notification frequencies anchored to?
- Resend logic is anchored to original deployment time of the notification

**Comment by hydawo** [comment]
@MMel099 

For Yale, let's have them do basic tests:
Upon registration
- Enroll, set enrollment date
- Record which surveys are initially available
- Take one copy of ALWAYS AVAILABLE and one copy of ABSOLUTE
- Record if ABSOLUTE survey repopulates, take it, count how many times it repopulates

Throughout enrollment
- Check that DAILY notification comes in every day at 10:15AM and RELATIVE notification every day at 11AM
- On some days, take DAILY survey before RELATIVE survey comes in (between 10:15AM and 11AM). Then take RELATIVE survey after it is sent at 11AM
- On some days, take both surveys in the evening
- On some days, don't take either survey

**Comment by biblicabeebli** [comment]
@MMel099 @hydawo @jprince127 I've got a server-side update in the works for testing, please keep track of items that do and don't work correctly so we can check for regressions when I next update the server.
- (tech detail - there is a known bug where absolute schedules get regenerated, working on that bug has lead to some larger rework to prevent that class of issue occurring.)

**Comment by biblicabeebli** [comment]
And let's add to the testing list - check resend behavior (and app behavior) if you _clear_ the notfication and then proceed with various actions (opening the app, dismissing the app from the app switcher, reboot the phone...)

**Comment by MMel099** [comment]
@biblicabeebli   Testing feedback TLDR: Everything behaving as normal **EXCEPT one day with missing surveys** 

Missing surveys in question happened on Friday 11/08. I went to take the surveys around 6pm and noticed that out of the three surveys that I expected to see, only 1 was there. Always available was there while the relative and daily were not. I went in about two hours later and the relative survey appeared while the daily was still missing. My notification history confirms this - I received resend notifications until about 6pm when I opened the app for the first time.

This is on Hassan's beta study (Staging: GGcUUuyFVGvPwogTsGs4CeNR). My Beiwe id is [79awifdq](https://staging.beiwe.org/view_study/80/participant/79awifdq)

> What happens if you don’t take a survey?
> 
> * Test 1: don't take survey until 7 or 8pm

Notifications continue at the expect interval until the app is opened

> * Test 2: don't take survey at all for a day, pay attention to notification behavior on day after. Make sure there's only 1 copy of each daily/relative survey to take the day after you skip them

Notifications continue as expected. The next day, only one copy of each survey is available.

**Comment by jprince127** [comment]
Hello! I believe the survey resend behaved as expected for me.
> Do pushes and survey package download occur over data as well as wifi?
> * Test 1: take your phone off WiFi for a day, record behavior of notifications
I continued to get regular notifications without Wifi

> Does the device communicate with the server once a survey is complete?
> * Test 1: interact with a survey notification, make sure survey shows in queue, do not take and see if notifications continue
> * Test 2: interact with a survey notification, make sure survey shows in queue and start taking survey, do not complete
Under both of these, I no longer had survey notifications for the rest of the day

However, I did notice there is some strange scheduling with the notifications. We had survey's scheduled to go out at 10:15AM and 11AM, and I received notifications at: 10:18, 10:54, 11, 11:30, 11:36, 12, and 12:06 for at least 3 days. There seems to be a 3 minute offset in the beginning for the first survey, then surveys come in at 36/30 minute increments.

**Comment by biblicabeebli** [comment]
(Ok. I am mostly recovered from the plague. SitRep!)

I just finished (god willing) and deployed to our staging server my various extremely delayed updates to push notification scheduling.

### TLDR:
Sometimes a software component attacks a problem "in the wrong direction" - decisions made in the past fight against changes imposed in the future.  In building the resend feature that you have been testing, I identified various _other_ ~bugs, oversights, obscure-interactions-with the survey scheduling system.  Ideally you see no change at all - but I can finally make simple deductions about whether an issue is sourced from _the scheduling code_ or _the resend code_.

Uh.

Please let me know if things change, both good change things and bad change things.

(I will... review this thread soon... I have an inbox that is... on fire.)

**Comment by biblicabeebli** [comment]
@MMel099 
https://github.com/onnela-lab/beiwe-ios/issues/69#issuecomment-2474473825

>Missing surveys in question happened on Friday 11/08. I went to take the surveys around 6pm and noticed that out of the three surveys that I expected to see, only 1 was there. Always available was there while the relative and daily were not. I went in about two hours later and the relative survey appeared while the daily was still missing. My notification history confirms this - I received resend notifications until about 6pm when I opened the app for the first time.

Just want to confirm - there was a daily (weekly-style schedule) survey that you never saw / were unable to take?

**Comment by biblicabeebli** [comment]
@jprince127 
> However, I did notice there is some strange scheduling with the notifications. We had survey's scheduled to go out at 10:15AM and 11AM, and I received notifications at: 10:18, 10:54, 11, 11:30, 11:36, 12, and 12:06 for at least 3 days. There seems to be a 3 minute offset in the beginning for the first survey, then surveys come in at 36/30 minute increments.

I'll work on that, I forgot about that, it's usually I forgot to zero some like a datetime microseconds value to a time comparison gets weird.

**Comment by MMel099** [comment]
Hey @biblicabeebli 

It has not occurred since then. The notification is in the history and i pulled this screenshot which shows persistent resends up until 7pm at which point I must have opened the app.

**Comment by MMel099** [comment]
TLDR: One user reports going into Beiwe and not seeing the survey available, resend logic sent a new survey on the next iteration

Following up on this thread to document a missing survey from one of the Yale researchers on the staging Beta testing study.
Study: https://staging.beiwe.org/view_study/95
User: https://staging.beiwe.org/view_study/95/participant/qd4eimro
Device: iOS 18.1 running Beiwe 2024.43
Date: Nov 27th, 2024 - a diary survey should have been available at 9am (this would be 1 day after enrollment date) 
The user reported seeing the notification come in around 9am, however they did not interact with it and only opened the app a few hours later closer to 11am. Once in the app, there was no survey available. A manual resend was completed and the survey became visible. At the 3 hour mark, the resend logic kicked in and another notification was sent as well.

**Comment by biblicabeebli** [comment]
@MMel099 Bring this up with @hydawo at your meeting tomorrow if he hasn't chimed in.
That behavior is as designed, but highlights a detail I had totally overlooked - the similar names of the things.
Similar names, totally unrelated code - I don't why I didn't notice the naming could be confusing. 
(maybe because they are VERY unrelated. I don't think it was touched except for a very minor update.)

@hydawo the "best" fix here is for manual resends to link to automatic-resends such they restart the automatic-resend-periodicity for that participant.  If we are lucky then it is trivial?  We are not very lucky.  It might be best to delay that detail and get automatic resends out (this dev period is already heavily dragging), and then we improve it later.

**Comment by biblicabeebli** [comment]
App Build 2024.45 contains a breaking change. Older versions of the app will (I think) be stuck in an endless loop of resends because they no longer understand the format of the data the backend sends them.

Apologies, there was an oversight on my part for the last detail to get all the data where it needs to be for the app to report notification uuids.

**Comment by MMel099** [comment]
Another interesting issue has come out of Yale's beta testing. They had an instance where the survey came in and was available but once opened, there was no "next" button to proceed to the next question. Attaching a screenshot with the missing "next" button below.

<img width="300" alt="image" src="https://github.com/user-attachments/assets/b1637ddc-5830-4071-a11d-079a452e512e">


Separately, the Yale study has seen some instances where participants do not get surveys including manual pushes from the dashboard. I am wondering if you have any sense as to whether resends would resolve this?

Thank you!

**Comment by biblicabeebli** [comment]
@MMel099 can we get an ios version and a phone model on that device with the missing next button, it would help if I had a survey too.

**Comment by biblicabeebli** [comment]
> Separately, the Yale study has seen some instances where participants do not get surveys including manual pushes from the dashboard. I am wondering if you have any sense as to whether resends would resolve this?

Is this of people with the beta, or do you mean existing issues they are reporting that have this?  Are these lost notifications and no surveys to take, or no notification but there are surveys?  (ios 18 has new notification filtering so it is possible we are hitting that.)

**Comment by MMel099** [comment]
These are existing issues with the current public version. These seem to be lost notifications with no surveys to take. Most of the time when we hear about missing notifications/surveys, a manual resend works. However, this caught my eye because the manual push does not go through.

(Also, I edited my comment above with details of the missing next button)

**Comment by biblicabeebli** [comment]
OK - adding this comment after the fact and after like an hour - this might be a known issue - (but I'm done for the day):
- during development of date/time questions, when interacting with the conditional question display logic... which was a few years ago now...
- I discovered an issue where these buttons would [need to?] disappear _when navigating backwards from ...._
   - I think it was when the next a question that was currently displayed had conditional logic that depended on the previous question that you were going back to...
   - and/or/but maybe it was also... when .... going backwards skipped over a bunch of questions that were never displayed?
     - because the app state only considers the current question and doesn't know which question in the list it is jumping to, because they may themselves have conditional logic.
     - ooorrrr possibly _even more_ obscure, like ... it doesn't know until "too late" (After the BACK button get's pressed)
       - and then isn't "given the chance" to check which question will be displayed as a result of going backwards
       - so from the code's perspective the app "suddenly renders the question it has jumped to + those buttons without telling it what its doing"
       - so in order to correctly display/not-display a pressable SKIP  or  NEXT button on a unknown-but-potentially-required question.... it is forced to err towards hiding them
         - (and then it finally learns which question is currently displayed only after the user interacts with it, and that is in particular bad UX with Date questions because the "autofill" weirdly.)
- Yeeah.   It was something **super obscure** like that.
- BUUUUUUUUT I can't remember if it was ios or android.  🫠 ....
  - I know I posted it somewhere, possibly in our internal work issues, but could be on the android or ios... ug....
  - @MMel099 @jprince127 @hydawo I could use help tracking this one down.
    - the ios app store release notes say Jul 21, 2023 was when we released date/time questions for iOS
    - if this can be reproduced and shown that the app correctly displays the next buttons after interacting with it then this is user error, not release blocker.


Regarding:
<img width="300" alt="image" src="https://github.com/user-attachments/assets/b1637ddc-5830-4071-a11d-079a452e512e">

- Is it possible that this is a required question, and they need to change the value in order for the app to count it as selected and display the buttons?
  - iirc date/time questions cannot be set as required in the web interface due to issues like this across both android and ios.
    - but editing the question JSON directly can create such a question (probably)
    - or maybe if you just set a question as required and then changing the question type results in a required date/time - if so that is a bug on the backend, and.... that's fun but not a release blocker.


I need the survey that has that question, and I guess I need the participant ID.  (email a link to their origin post, I might need to talk with their tester.)

Assuming it's not the question above -- What happens when that survey on that device gets loaded now?  is the user blocked entirely from ever finishing that survey?
- If there is a tiny random chance where sometimes the survey transiently the doesn't show the button for a question, or if a back-forward can cause this, then we should proceed with the release and debug this later.  
- Buuuuuut if we can reproduce it ~easily then we need to stress test and it is going to be awful.


The rest of this is stream of consciousness brainstorming on how to even do anything for this one, no need to respond to anything below:

- If this is an app-side _code_ bug in the button/question rendering component of researchkit then...
  - it will be reproduceable regardless of the device.
  - and it will happen randomly or _seemingly_ randomly.
  - _usually the weird stuff_ happens under ... "high device load" conditions.
    - (I guess I can run a memory pressure test? this requires xcode)
    - I can check version updates to researchkit, the newest version cuts out a feature we use on the consent screen and I would have to write that entire flow from scratch
      - that would be a disaster, I don't think that would be a 2024 thing.

- if this app installation becomes perma-stuck with no rendered button  _OR_  This goes away only after receiving a notification for that same survey again  ....
  - then it is a bug where the specific local copy of the survey in live memory is.... somehow at fault.
  - _ALSO_  if this happens but then goes away after dismissing the app from the task switcher and reopening, then it ... involves the local database too? 

- If this happens on only that question type.... ok we narrowed down the bug, cool.

- if this happens only on required questions - 

- if this can be caught on an always-available or trigger-on-first-download survey, and the behavior is different ... that ... would be interesting, but not unexpected.

- The only way to test the entire render process for fully disappearing and reloading the survey is by sending a push notification.  Maybe.
  - TODO: when a survey notification is received for closed, _partially-completed_, survey - does the survey progress reset?
    - if yes, then receiving a survey notification probably causes a full survey reload.
    - if not, then it just marks the survey as visible..... - I would have to read a bunch of code.

**Comment by hydawo** [comment]
Hi @biblicabeebli here is some information on my beta testing ID that had multiple surveys sent/completed on 12/3/24

[m1gz9ib3](https://staging.beiwe.org/view_study/80/participant/m1gz9ib3) in the [Hassan_Beta_Test_2_13_23](https://staging.beiwe.org/view_study/80) on Staging

I took surveys normally for quite a while (from registration to 11/25/24). I stopped for ~1 week and resumed surveys on 12/2/24. On 12/3/24, I received the daily Diary survey 3 times. I was able to take it each time and shortly thereafter (within an hour) another copy of the survey would be received and I would take it. 

I've confirmed in the output data that there are 3 answer files and 3 timings files uploaded on 12/3/24. 

I've now moved to another beta testing ID in the Yale beta study on staging, so this ID is no longer active. Let me know if I can provide more details. Thank you!

**Comment by hydawo** [comment]
@MMel099 @jprince127 @QianJingyun make sure you are running the latest TestFlight build (2024.45) and registered to the [Yale Beta Study](https://staging.beiwe.org/view_study/95)

Here are 2 tests we need to perform this week; we'll discuss further in our RA meeting:

Test 1: Skip a day of surveys → wait for next day’s survey → take next day’s survey and monitor if another survey is deployed to you (before next scheduled time)
Test 2: Receive a survey → wait 1 hour → manually resend survey to yourself → check how many surveys you have available → take them (hopefully only 1 copy) → monitor if another survey is deployed to you (before next scheduled time)

**Comment by hydawo** [comment]
To test this, let's do the following:

**Test 1: I**n the Diary survey, answer question 1, click next and move to question 2 (do not answer question 2), go back to question 1, is Next button present?

**Test 2:** In the Diary survey, answer question 1, click next and move to question 2, answer question 2, go back to question 1, is Next button present?

> We need to identify if this bug: <img alt="image" width="300" src="https://private-user-images.githubusercontent.com/63422396/393398300-b1637ddc-5830-4071-a11d-079a452e512e.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4NjU0OTQsIm5iZiI6MTczMzg2NTE5NCwicGF0aCI6Ii82MzQyMjM5Ni8zOTMzOTgzMDAtYjE2MzdkZGMtNTgzMC00MDcxLWExMWQtMDc5YTQ1MmU1MTJlLlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEwVDIxMTMxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwNzJhZTZkMzYyNmNjZDNlM2ExMTk4OWI3ZjMzZmRjNzdiNjUxYjU3ZDE3N2IxY2U2ZmVlNThmYjZjODc0OTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.KZcMTp34U1U_x5cKDLAXBA5dIXfFtEahapH7IFUTzUI"> is actually this issue:
> 
> * If you clear out the answers for a required question (this cannot be done with all question types, like radio buttons and date/time) and then go back 1 question to another [answered] required question, that previous question will be displayed without a next button until you refresh the answer by interacting with the question.
> 
> (this is inside a collection of minor ui weirdnesses, which are collected in issue # 199 over on an old tracking issue in our internal tracking repo. (We should break those out into tracking issues in their respective repo)

**Comment by QianJingyun** [comment]
Steps to replicate the bug I found on meeting Dec 10 around 16:30.
1. After login, in the Study, Active Surveys, with 'Survey - Welcome Survey', [I'm not sure what was the status (which currently is incomplete)]
![IMG_6142](https://github.com/user-attachments/assets/1f8ec88b-f5c8-469d-bf0c-09a8722647c1)
2. Click on 'Survey - Welcome Survey'. I see what's in the screenshot below. Available buttons: 'Get Started' and 'Cancel'
![IMG_6143](https://github.com/user-attachments/assets/788605af-b2f8-4ee5-8f06-e06c98de26ea)
3. Click on 'Get Started'.  I see what's in the screenshot below. [Available buttons: 'Back' and 'Cancel'. I think when the first time I entered this survey, I could see a next button]
![IMG_6144](https://github.com/user-attachments/assets/5665eafd-e50b-448b-9c44-2b279470c95b)
4. [I answered the question the first time by scrolling the bar from 0 to the middle.]
5. (For this step, I'm not sure what I did in the first time) Click on 'Back'. I see what's similar to the screenshot below. [I cannot take a screen shot right now, because it crashes and quit unexpectedly]
![IMG_6143](https://github.com/user-attachments/assets/136245c2-98fc-415e-a01d-09cdaff3ffcb)
7. Click on 'Cancel'.  I see what's in the screenshot below.
![IMG_6141](https://github.com/user-attachments/assets/c7fe6c5e-c042-4119-b4d8-715d70bc4b30)
8. Click on 'End Task'. I see what's in the screenshot below.
![IMG_6142 2](https://github.com/user-attachments/assets/bac794ae-df06-43f7-85e7-3be235db4b49)

**Comment by biblicabeebli** [comment]
I've got a possible fix for the zombie survey notifications deployed to staging.

TODO: clean up this thread a bit...

**Comment by biblicabeebli** [comment]
@biblicabeebli testing if I can send ANY notification to myself....

**Comment by jprince127** [comment]
Hello!

Here are the results of my testing:

> Test 1: Skip a day of surveys → wait for next day’s survey → take next day’s survey and monitor if another survey is deployed to you (before next scheduled time)
I skipped the survey on Wednesday (12/11) and waited for the next day's survey and there was not another survey that was deployed to me before the next scheduled time

> Test 1: In the Diary survey, answer question 1, click next and move to question 2 (do not answer question 2), go back to question 1, is Next button present?
Yes, the next button is present

> Test 2: In the Diary survey, answer question 1, click next and move to question 2, answer question 2, go back to question 1, is Next button present?
Yes, the next button is present

However, I did manage to trigger a survey answer bug, where I answered question 1, then answered question 2, progressed to the third question, then went back to the first question, and hit cancel. Then, when I reopened the survey, there was no longer the back button, and also on the THIRD question there was no longer an option to progress past it. Attached is a screenshot (no "Next" or "Back" button)

<img src="https://github.com/user-attachments/assets/6dad05df-25c0-4dc0-a6d4-011043a3af26" width="100" />

**Comment by QianJingyun** [comment]
Hello!

Here are the results of my testing:

Test 1: Skip a day of surveys → wait for next day’s survey → take next day’s survey and monitor if another survey is deployed to you (before next scheduled time)

I skipped the 'Diary' survey on Wednesday (12/11) and both the 'Diary' and 'Audio Survey - The Caterpillar Passage' survey on Thursday (12/12) between 9:15 AM and 9:21 AM. There was not another survey that was deployed to me, at least not until (12/13 3:00 AM). 


Test 1: In the Diary survey, answer question 1, click next and move to question 2 (do not answer question 2), go back to question 1, is Next button present?

Yes, the next button is present.

Test 2: In the Diary survey, answer question 1, click next and move to question 2, answer question 2, go back to question 1, is Next button present?

Yes, the next button is present.

I received notification ' Beiwe may not be running correctly, please open the Beiwe app.' This is probably because my phone closed the app back stage. But this did not affect me receiving notification nor finishing the survey.

**Comment by biblicabeebli** [comment]
Development of this version of the Beiwe app is completed, the app has been released (rolling release, over the next 7 days) as build 2024.48.  If you encounter bugs referenced here on the released app please create a new issue referencing this issue's appropriate comment..

Thank you everyone !

**Comment by hydawo** [comment]
@biblicabeebli reopening this issue, @MMel099 please add recent feedback from Yale regarding resend logic continuing to run past the next instance of a survey (i.e. Survey 1 and Survey 2 are the same survey "Diary". Survey 1's logic does not stop when Survey 2 is initially sent). Please add a screenshot of a participant's page showing this error. 

Thank you!

**Comment by MMel099** [comment]
Yale pointed out to us what Hassan described above - survey resends seem to continue indefinitely if the user does not check in for several days. This becomes especially problematic when a new survey is deployed (and then an additional schedule of resends commences). Ideally, a new survey should truncate any resends from previous surveys otherwise the resends begin to stack. The screenshot below shows a snippet of the resend schedule for [j96jck8o](https://studies.beiwe.org/view_study/952/participant/j96jck8o). The diary that was deployed at 9am on 01/22 and continued resending even at 5pm on 01/23. We are hoping for the 9am deployment on 01/23 to automatically disable any resends from previous days. 

![Image](https://github.com/user-attachments/assets/fd3e96d0-3612-44e9-9099-8b52fa8d9ea1)

They also shared this screenshot yesterday from the same participant. It looks like the structure of the notification page being organized by survey deployment begins to break down when several days of resends begin to stack. Here's the [link](https://studies.beiwe.org/view_study/952/participant/64ehsy9k/notification_history) to the notification page of this user.

![Image](https://github.com/user-attachments/assets/7569e262-fbaf-4bf9-abfb-6cff31b6334e)

**Comment by biblicabeebli** [comment]
@hydawo @MMel099 this issue was for release issue tracking and device testing of the beta build for 2.5.1/2.5.2

The items recently added are backend issues, not app issues.

I pushed to staging last night the first deploy of a hopeful fix, which I have not written up a description for yet but has been communicated to Hassan. (not the best order of operations, I admit, but y'know time in the day and all that)

Please create new issues on the beiwe-backend if there is not one already, one for each separate issue as you understand them, and I will coalesce the into a single issue based on ~development relatedness.  Please reference the comments in this issue as their origin post, that will automatically create a reference trail in both issues. When you are finished please re-close this issue.

don't worry if you haven't seen a new issue from me yet about the pushed update, I will handle mess like that.

**Comment by biblicabeebli** [comment]
Issue is old enough now to be closed.
